### PR TITLE
Create shipping address view

### LINKED ISSUE #4894: [Order Editing] Create Shipping Address form view
We should be capable of doing the following with an dedicated Shipping Address view:
- Configure the `Use as Billing Address` toggle
- Trigger optimistic update request when hitting the `done` button
- Remove the `email` field from the view
- Configure the Address section label with the `Shipping` string

Depends on #4886 and #4978 

Android-order-details-edit-billinginfo

### LINKED ISSUE #4978: [Order Editing] Create generic override points to enable Base address form to be extended
One of the main objectives of creating the Base Address Form is to enable usage for general-purpose Address editing, which can be useful in multiple parts of the app, but for Order Editing is key for Shipping and Billing Address data, to achieve that we need to leave abstract implementation points to enable a more customizable usage of the Fragment.

Identified points:
- Fragment title
- `Use as` toggle behavior, text, function call and visibility
- Address section text
- Done button function call

### PR DESCRIPTION
**Pull request by ThomazFB**
Summary
==========
Fixes issue #4894 and #4978 by creating abstraction points in `BaseAddressEditingFragment` to allow it be extended and adjust its behavior, but also implementing these abstractions for the `ShippingAddressEditingFragment` allowing us to properly edit all expected Shipping Address fields.

Please note that this PR doesn't implement anything on the Billing Address editing view yet, this will be done in upcoming PRs.

Screenshots
==========
| Editing the Shipping Address info  | Shipping Address updated without checking the `use as billing address` toggle | Both order addresses updated when checking the `use as billing address` toggle
| ------------- | ------------- | ------------- |
| Screenshot_20211021_212142 | Screenshot_20211021_212217 | Screenshot_20211021_212232 |

How to Test
==========
## Scenario 1
1. Go to the Order Details of any existent Order
2. Go to the Shipping Address section and click to edit or add one
3. Edit any field and verify if the `done` button is displaying as expected
4. Verify that clicking on the `done` button effectively updates the Shipping Address using the optimistic update strategy
5. Now back at the Order Details, verify if the information is displayed correctly at the view and matches with the Shipping information displayed just the same at the site.

## Scenario 2
1. Go to the Order Details of any existent Order
2. Go to the Shipping Address section and click to edit or add one
3. Verify that the `use as Billing Address` toggle is deactivated if Shipping and Billing Addresses are different from one another
4. Edit any field and verify if the `done` button is displaying as expected
5. Verify that when editing the Shipping Address, if the `use as Billing Address` toggle is activated, the Billing address will be updated too, but keeping the original email there
6. After submitting the Address to both Shipping and Billing, verify if the Shipping Address Editing view displays the `use as Billing Address` toggle already activated.

Update release notes:

- [X] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Comment by nbradbury** [comment]
@ThomazFB There appears to be a FluxC database migration issue here.

```
java.lang.RuntimeException: Unable to create application com.woocommerce.android.WooCommerceDebug:
 android.database.sqlite.SQLiteException: duplicate column name: SHIPPING_PHONE
 (code 1 SQLITE_ERROR): , while compiling: ALTER TABLE WCOrderModel
```

**Comment by ThomazFB** [comment]
@nbradbury could that be a result of the correction I made to the DB version on [this PR](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2159)? This is the first PR to reference FluxC after that change, and our Woo app was already pointing to the FluxC version with the incorrect DB version number, so there's a chance that now that we have the correct version, the migration is trying to once again create the same column, I think this could be the case of reinstalling the app since the Woo app on develop is currently in an invalid DB version state.

**Comment by nbradbury** [comment]
@ThomazFB I noticed that if you toggle "Use as billing address," the Done button doesn't show up. Maybe you want to handle that in a separate PR?

Screenshot_20211022_103401

**Comment by ThomazFB** [comment]
>I noticed that if you toggle "Use as billing address," the Done button doesn't show up. Maybe you want to handle that in a separate PR?

@nbradbury, thanks for highlighting this! There's a current ongoing discussion on the iOS side about triggering the `done` button when changing the toggle not being the better approach since we can't track this toggle state through the API. So I intentionally just replicated the iOS behavior for the `use as` toggle right now.

If we implement this behavior, we may create some weird scenarios like just unchecking the toggle when the Shipping and Billing are already identical: There will be nothing to send as an update to the site in this case, and since there's no field to control the toggle state in the API, this toggle state is controlled by checking both models data equality, so the toggle will come back checked once the user reenters the View again.

I suggest creating a separate issue to track what we want to do with the toggle interaction, WDYT?

Btw, I've merged your PR with the Country/State data source into this one to solve some conflicts, so it may be worthy of testing the UI again.

**Comment by nbradbury** [comment]
> I suggest creating a separate issue to track what we want to do with the toggle interaction, WDYT?

Sounds good to me. And we may want to reconsider even having this toggle, because it's a bit odd. Perhaps the billing address screen could have a "Copy from shipping address" button?

**Comment by ThomazFB** [comment]
>Sounds good to me. And we may want to reconsider even having this toggle, because it's a bit odd. Perhaps the billing address screen could have a "Copy from shipping address" button?

This sounds like a much better idea, Nick!  Thanks! I think the Woo Core does just like that, and it makes more sense this as an action instead of a state that doesn't exist and needs to be artificially controlled! I'll place this suggestion on the issue and ping everyone to discuss what we should do!

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/details/editing/OrderEditingViewModel.kt [hunk 1]
```diff
@@ -61,29 +61,43 @@ class OrderEditingViewModel @Inject constructor(
...
+        if (viewState.useAsOtherAddressIsChecked == true) {
+            sendReplicateShippingAndBillingAddressesWith(updatedShippingAddress)
         } else {
-            false
-        }
+            orderEditingRepository.updateOrderAddress(
+                order.localId,
+                updatedShippingAddress.toShippingAddressModel()
+            )
+        }.collect()
```
Comment (nbradbury) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/details/editing/OrderEditingViewModel.kt:78]: This doesn't have to be addressed in this PR, but I wanted to point out that `collect()` only tracks the `ORDER_EDIT_CUSTOMER_NOTE` event, so we need to add tracking for the addresses.

Comment (nbradbury) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/details/editing/OrderEditingViewModel.kt:78]: @ThomazFB I opened #5062 to address this.

**Review by nbradbury — APPROVED** [review]
Looks great!

**Pull request merged**